nemo/collections/asr/parts/numba/rnnt_loss/utils/cuda_utils/gpu_rnnt_kernel.py

```diff
@@ -875,3 +877,500 @@ def compute_multiblank_grad_kernel(
+
+            if t + durations[idx] == T and idx >= 1 and u == U - 1:  # for blank as the last symbol
+                grad -= math.exp(alphas[col] + logpk_blank - logll[mb])
+
+            grad = grad * math.exp(duration_acts[col * num_durations + idx])
+            duration_grads[col * num_durations + idx] = grad
+
+        # For cuda kernels, maximum number of threads per block is limited to some value.
+        # However, it may be the case that vocabulary size is larger than this limit
+        # To work around this, an arbitrary thread buffer size is chosen such that,
+        # 1) each element within the thread pool operates independently of the other
+        # 2) An inner while loop moves the index of each buffer element by the size of the buffer itself,
+        #    such that all elements of the vocabulary size are covered in (V + 1 // thread_buffer) number of steps.
+        # As such, each thread will perform the while loop at least (V + 1 // thread_buffer) number of times
+        while idx < alphabet_size:
+            # remember, `col` represents the tri-index [b, t, u]
+            # therefore; logpk = denom[b, t, u] + acts[b, t, u, v]
+            logpk = denom[col] + acts[col * alphabet_size + idx]  # - sigma
+            # initialize the grad of the sample acts[b, t, u, v]
+            grad = math.exp(alphas[col] + betas[col] + logpk - logll[mb])  # * math.exp(sigma)
+
+            # If FastEmit regularization is enabled, calculate the gradeint of probability of predicting the next label
+            # at the current timestep.
+            # The formula for this is Equation 9 in https://arxiv.org/abs/2010.11148, multiplied by the log probability
+            # of the current step (t, u), normalized by the total log likelihood.
+            # Once the gradient has been calculated, scale it by `fastemit_lambda`, as in Equation 10.
+            if fastemit_lambda > 0.0 and u < U - 1:
+                fastemit_grad = 0.0
+
+                for i in range(0, num_durations):
+                    if t + durations[i] < T:
+                        fastemit_grad += fastemit_lambda * math.exp(
+                            alphas[col]  # alphas(t, u)
+                            + (denom[col] + acts[col * alphabet_size + labels[u]])
+                            + duration_acts[col * num_durations + i]
+                            + betas[col + 1 + durations[i] * maxU]  # betas(t, u+1)
+                            + logpk  # log Pr(k|t, u)
+                            - sigma  # y_hat(t, u)
+                            - logll[mb]  # total log likelihood for normalization
+                        )
+            else:
+                fastemit_grad = 0.0
+
+            # Update the gradient of act[b, t, u, v] with the gradient from FastEmit regularization
+            grad = grad + fastemit_grad
+
+            # // grad to last blank transition
+            # grad[b, T-1, U-1, v=blank] -= exp(alphas[b, t, u) + logpk - logll[b])
```

Comment:
The RNNT comments need to be updated to accommodate TDTs duration

Reply:
Done.